### PR TITLE
[2D][TP] Add DDP TP integration and unit test

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103604

Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

### CODE REVIEW

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46731700).

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -1166,9 +1166,26 @@ def __setstate__(self, state):
             self.logger._set_static_graph()
 
     def _build_params_for_reducer(self):
+        from torch.distributed._tensor import DTensor
+        import functools
+        def _dtensor_ext(param_name, param):
+            def tensor_hook(param, grad):
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:1172]: what's the ordering of these hooks with respect to the Reducer installed hooks? How do we guarantee this ordering?

Comment (fduwjj) [reply torch/nn/parallel/distributed.py:1172]: This is not needed, actually. I haven't cleaned it up. I need to polish this PR a little bit more.

File: test/distributed/tensor/parallel/test_tp_ddp_2d.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    for item in module_list:
+        parent_module, module_path, t = item
+        if hasattr(parent_module, module_path):
+            delattr(parent_module, module_path)
+        setattr(parent_module, module_path, t)
+
+
+    twod_model = DDP(twod_model, process_group=dp_pg, find_unused_parameters=True)
+
+    # We also need a pre-forward hook to ensure gradients get propagated back.
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_tp_ddp_2d.py:126]: yes. it's called every forward.

File: test/distributed/tensor/parallel/test_tp_ddp_2d.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    dp_pg = twod_mesh.get_dim_groups()[0]
+
+    # Create Input
+    twod_model = parallelize_module(twod_model, twod_mesh, PairwiseParallel(), tp_mesh_dim=1)
+    module_list = []
+
+    # We need to find a way for DDP to replicate local tensor of DTensor.
+    for name, param in twod_model.named_parameters():
+        if isinstance(param, DTensor):
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/tensor/parallel/test_tp_ddp_2d.py:99]: yes. Not need.

**Comment by fduwjj** [comment]
Cannot ghstack import so created a new polished one in https://github.com/pytorch/pytorch/pull/106583.

**Pull request closed without merging**